rfcs/0006-transaction.md

```diff
@@ -0,0 +1,300 @@
+txn.commit().await?;
+```
+
+### Conflict Checking
+
+The approach we choose on conflict checking is expected to be more similar to Badger's global `Oracle` approach, since we hope to provide SSI in the transaction feature, and it's less prone to unnecessarily abort the transaction with limited MemTable size.
+
+Below is the data structure of the global `Oracle` and the transaction state:
+
+```rust
+struct Oracle {
+    next_seq: AtomicU64,
+    recent_committed_txns: Deque<Arc<TransactionState>>,
+}
+
+struct TransactionState {
+    started_seq: u64,
+    write_keys: HashSet<Bytes>,
+    read_keys: Vec<Range<Bytes>>,
+    committed_seq: Option<u64>
+}
+```
+
+`Oracle` is considered a global singleton in the DB, and it tracks the recent committed transactions in the `recent_committed_txns` deque. When a transaction commits, it should push itself into the `recent_committed_txns` deque, and the entries in `recent_committed_txns` can be GCed after they are no longer needed. Also, when a `TransactionState` is tracked in the `recent_committed_txns`, only the write keys need to be tracked, while `read_keys` should always be empty.
+
+Besides the started sequence number of the transaction, we also need to track the sequence number of the transaction committed. During the execution of a transaction, some other transactions might commit, and the current transaction should check the conflicts with these committed transactions.
+
+The conflict check simply checks whether the Read & Write keys of the current transaction are overlapped with the Write keys from related recent committed transactions.
+
+Given an execution history of transactions like:
+
+```
+txn1: [seq:97] <--- ---- ---> [seq:99]
+txn2:     [seq:98]  <--- ---- ---- ---- ---- ---- ---> [seq:105]
+txn3:                         [seq:102] <--- ---> [seq:104]
+txn4:                    [seq:101] <--- ---- ---- ---- ---> [seq:106]
+```
+
+Let's say the current transaction is `txn4`, it should check conflicts with `txn2` and `txn3` during the commit, because `txn2` and `txn3` are committed after `txn4` started. During the conflict checking, `txn4`'s Read & Write keys should not conflict with the Write keys of `txn2` and `txn3`.
+
+It does not need to check conflicts with `txn1`, because `txn1` is committed before `txn4` started. This could tell us when to GC the entry in the `recent_committed_txns` deque: when ALL the running transactions' sequence number are bigger than the committed sequence number of the entry, then the entry could be GCed.
+
+Please note that `read_keys` are represented as a set of `Range<Bytes>`. This allows us to efficiently track ranges of keys that were read during the transaction, and ensure Serializable Snapshot Isolation (SSI) on range conflicts. For example, if transaction A reads a range `["key01", "key10"]` and transaction B writes to `"key05"` which falls within that range, we can detect this conflict and abort one of the transactions to maintain serializability.
```

Comment:
for range scans should we record the entire range or just the last key returned from the iterator (or the full range if iterator is exhausted)? for example I suspect that many range scans will return early once they find a first match (e.g. "find me first key >= val"). If that's the case we only need to record [val, returned_key] in the range

also nice find, I wouldn't have thought about this!

Reply:
i guess we can only track the real scanned key range in the iterator instead of the full range. if the iterator returns early, we can take the last key it scanned into this range for conflict checking.